File: contest/ICPC/2018/北美洲/北美洲中北部/data/maximumsubarrays/submissions/accepted/bryce.java
import java.util.*;

public class bryce {

    static long INF = 1000000000000000000L;

    public static void main(String[] args) {
        Scanner in = new Scanner(System.in);
        int n, k;
        n = in.nextInt();
        k = in.nextInt();

        long[] A = new long[n];
        for (int i = 0; i < n; ++i) {
            A[i] = in.nextInt();
        }

        long[][] Best = new long[n][k+1];
        for (long[] arr : Best) {
            Arrays.fill(arr, -INF);
        }
        long[][] BestInclude = new long[n][k+1];
        for (long[] arr : BestInclude) {
            Arrays.fill(arr, -INF);
        }

        for (int i = 0; i < n; ++i) {
            Best[i][0] = 0;
        }
        Best[0][1] = BestInclude[0][1] = A[0];

        for (int i = 1; i < n; ++i) {
            for (int j = 1; j <= k; ++j) {
                BestInclude[i][j] = Math.max(BestInclude[i-1][j], Best[i-1][j-1]) + A[i];
                Best[i][j] = Math.max(Best[i-1][j], BestInclude[i][j]);
            }
        }

        System.out.println(Best[n-1][k]);
    }

}

File: contest/ICPC/2018/北美洲/北美洲中北部/data/maximumsubarrays/submissions/accepted/bryce.cpp
#include <iostream>
#include <vector>
#include <cmath>

using namespace std;

typedef long long ll;
typedef vector<ll> vi;
typedef vector<vi> vvi;

static const ll INF = 0xffffffffll * 100000ll;

int main() {
    ll n, k;
    cin >> n >> k;
    vi A(n);
    for (ll i = 0; i < n; ++i) {
        cin >> A[i];
    }

    vvi Best(n, vi(k+1, -INF));
    vvi BestInclude(n, vi(k+1, -INF));

    for (ll i = 0; i < n; ++i) {
        Best[i][0] = 0;
    }
    Best[0][1] = BestInclude[0][1] = A[0];

    for (ll i = 1; i < n; ++i) {
        for (ll j = 1; j <= k; ++j) {
            BestInclude[i][j] = max(BestInclude[i-1][j], Best[i-1][j-1]) + A[i];
            Best[i][j] = max(Best[i-1][j], BestInclude[i][j]);
        }
    }

    cout << Best[n-1][k] << endl;
}
